// playcanvas/build/playcanvas.prf/src/framework/lightmapper/render-pass-lightmapper.js
import { RenderPass } from '../../platform/graphics/render-pass.js';
import { SHADER_FORWARD } from '../../scene/constants.js';

class RenderPassLightmapper extends RenderPass {
		constructor(device, renderer, camera, worldClusters, receivers, lightArray){
				super(device), this.viewBindGroups = [];
				this.renderer = renderer;
				this.camera = camera;
				this.worldClusters = worldClusters;
				this.receivers = receivers;
				this.lightArray = lightArray;
		}
		destroy() {
				this.viewBindGroups.forEach((bg)=>{
						bg.defaultUniformBuffer.destroy();
						bg.destroy();
				});
				this.viewBindGroups.length = 0;
		}
		execute() {
				const device = this.device;
				const { renderer, camera, receivers, renderTarget, worldClusters, lightArray } = this;
				if (device.supportsUniformBuffers && !renderer.viewUniformFormat) {
						renderer.initViewBindGroupFormat(renderer.scene.clusteredLightingEnabled);
				}
				renderer.renderForwardLayer(camera, renderTarget, null, undefined, SHADER_FORWARD, this.viewBindGroups, {
						meshInstances: receivers,
						splitLights: lightArray,
						lightClusters: worldClusters
				});
		}
}

export { RenderPassLightmapper };
